File: packages/sdkwork-utils-go/validation.go
package utils

import "regexp"
import "strings"

var (
	emailRe = regexp.MustCompile(`^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$`)
	uuidRe  = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$`)
	urlRe   = regexp.MustCompile(`^https?://[^\s/$.?#].[^\s]*$`)
	ipv4Re  = regexp.MustCompile(`^(25[0-5]|2[0-4]\d|1?\d?\d)(\.(25[0-5]|2[0-4]\d|1?\d?\d)){3}$`)
	e164Re  = regexp.MustCompile(`^\+[1-9]\d{1,14}$`)
)

func IsEmail(value string) bool {
	return emailRe.MatchString(Trim(value))
}

func IsUUID(value string) bool {
	return uuidRe.MatchString(Trim(value))
}

func IsURL(value string) bool {
	return urlRe.MatchString(Trim(value))
}

func IsNumeric(value string) bool {
	_, ok := ParseNumber(value)
	return ok
}

func IsIPv4(value string) bool {
	return ipv4Re.MatchString(Trim(value))
}

func isIPv6Shape(value string) bool {
	if value == "" {
		return false
	}
	for _, ch := range value {
		if (ch < '0' || ch > '9') && (ch < 'a' || ch > 'f') && (ch < 'A' || ch > 'F') && ch != ':' {
			return false
		}
	}
	if strings.Count(value, "::") > 1 {
		return false
	}
	isValidPart := func(part string) bool {
		return len(part) > 0 && len(part) <= 4
	}
	if strings.Contains(value, "::") {
		parts := strings.SplitN(value, "::", 2)
		leftParts := strings.FieldsFunc(parts[0], func(r rune) bool { return r == ':' })
		rightParts := strings.FieldsFunc(parts[1], func(r rune) bool { return r == ':' })
		for _, part := range append(leftParts, rightParts...) {
			if part != "" && !isValidPart(part) {
				return false
			}
		}
		leftCount := 0
		for _, part := range leftParts {
			if part != "" {
				leftCount++
			}
		}
		rightCount := 0
		for _, part := range rightParts {
			if part != "" {
				rightCount++
			}
		}
		return leftCount+rightCount < 8
	}
	segments := strings.Split(value, ":")
	if len(segments) != 8 {
		return false
	}
	for _, part := range segments {
		if !isValidPart(part) {
			return false
		}
	}
	return true
}

func IsIPv6(value string) bool {
	return isIPv6Shape(Trim(value))
}

func IsPhoneE164(value string) bool {
	return e164Re.MatchString(Trim(value))
}

File: packages/sdkwork-utils-typescript/src/validation.ts
const EMAIL_RE = /^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$/;
const UUID_RE =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$/;
const URL_RE = /^https?:\/\/[^\s/$.?#].[^\s]*$/;
const IPV4_RE =
  /^(25[0-5]|2[0-4]\d|1?\d?\d)(\.(25[0-5]|2[0-4]\d|1?\d?\d)){3}$/;
const E164_RE = /^\+[1-9]\d{1,14}$/;

function isIpv6Shape(value: string): boolean {
  if (!value || !/^[0-9a-fA-F:]+$/.test(value)) {
    return false;
  }
  if ((value.match(/::/g) ?? []).length > 1) {
    return false;
  }
  const isValidPart = (part: string) => part.length > 0 && part.length <= 4;
  if (value.includes("::")) {
    const [left, right] = value.split("::");
    const leftParts = left ? left.split(":").filter(Boolean) : [];
    const rightParts = right ? right.split(":").filter(Boolean) : [];
    if (!leftParts.every(isValidPart) || !rightParts.every(isValidPart)) {
      return false;
    }
    return leftParts.length + rightParts.length < 8;
  }
  const parts = value.split(":");
  return parts.length === 8 && parts.every(isValidPart);
}

export function isEmail(value: string): boolean {
  return EMAIL_RE.test(value.trim());
}

export function isUuid(value: string): boolean {
  return UUID_RE.test(value.trim());
}

export function isUrl(value: string): boolean {
  return URL_RE.test(value.trim());
}

export function isNumeric(value: string): boolean {
  return Number.isFinite(Number.parseFloat(value.trim()));
}

export function isIpv4(value: string): boolean {
  return IPV4_RE.test(value.trim());
}

export function isIpv6(value: string): boolean {
  return isIpv6Shape(value.trim());
}

export function isPhoneE164(value: string): boolean {
  return E164_RE.test(value.trim());
}
